### DISCUSSION TITLE
storage.remove not removing item

### DISCUSSION THREAD

**Discussion by jaredcosulich** [supabase/supabase#4133] [Questions] [answered]
Per the documentation found here:

https://supabase.com/docs/reference/javascript/storage-from-remove

I'm currently calling:

const { data, error } = await supabase
  .storage
  .from('avatars')
  .remove(['folder/avatar1.png'])
  
with the correct bucket and file path for my app. This seems to be successful, returning:

{ "data": [], "error": null }

but when I visit the supabase website the file is still there. I imagine this is due to the lack of a policy. Is there documentation on policies somewhere that provides more examples than the few templates available?

**Comment by ltanady** [comment, marked as answer]
You are probably missing the policy for **DELETE**. Here's an example SQL expression for **DELETE** policy that works:

`((bucket_id = 'test-bucket'::text) AND (storage.extension(name) = 'png'::text) AND (lower((storage.foldername(name))[1]) = 'test-folder'::text))`

You can go to **Storage > Policies** to create policies for your bucket, 
<img width="1876" alt="Screenshot 2021-11-30 at 14 09 50" src="https://user-images.githubusercontent.com/888245/144002411-2a5476fb-a4a4-485b-b259-74b79dfb4c46.png">

For this case you need to use **DELETE** for allowed operations and provide the above SQL expression:
<img width="961" alt="Screenshot 2021-11-30 at 14 16 06" src="https://user-images.githubusercontent.com/888245/144002690-61dea59a-8822-4871-af7d-2f0030c09ef6.png">

Depending on who you want to allow to access, here are more examples:
https://supabase.com/docs/guides/storage#allow-individual-access-to-a-file

**Reply by tanifort** [reply]
This, somehow, still doesn't produce the appropriate outcome from me. I have added a policy for every user possible (authenticated, anon etc…) 
. I also made the bucket public, but I still receive an empty array as a result.

**Reply by GaryAustin1** [reply]
To delete you need a USING policy for delete and a USING policy for select.  Supabase does a read as part of delete.
Public just means you can get file at anytime with a fixed URL (it goes right to the "file" system.)  It does not change need for policy to upload, or delete.

**Reply by EugenVolosciuc** [reply]
After adding USING delete and select policies, I am still not able to remove files from my bucket.

These are the SQL queries I ran to create the policies:
![resume policies](https://user-images.githubusercontent.com/44806687/163143921-72f5725c-001f-4656-a209-03838a13a9c2.PNG)
Is there anything I might've missed?
I tried it without the owner check and it didn't work anyway.

It's strange that I'm receiving a 200 after removing the file, although it is clearly still in the bucket.

**Reply by GaryAustin1** [reply]
@EugenVolosciuc Are you sure the file is not just still cached?  Check in the UI and see if missing.

**Reply by EugenVolosciuc** [reply]
The file is still there.

This is the order of requests that I'm making:
`Remove file` -> [`Upload new file` and `Update user` in parallel] (I tried running them one after the other but it's the same result)
![uploadhandler](https://user-images.githubusercontent.com/44806687/163216749-6e35b98d-319a-4f95-bbb4-550cea91923c.PNG)

Here is the Supabase handling of these service methods:
![userservice](https://user-images.githubusercontent.com/44806687/163217245-6fb32368-1325-45e3-bcde-00e776ce5f7f.PNG)

Here is the response for the Delete request:
![deleterequest1](https://user-images.githubusercontent.com/44806687/163217860-a0b24ddd-c117-418e-ba96-413999301463.PNG)
![deleterequest2](https://user-images.githubusercontent.com/44806687/163217867-d9fdccf8-259f-4b06-88ee-3280c3a086be.PNG)
![deleterequest3](https://user-images.githubusercontent.com/44806687/163217871-0b361bbb-23a3-4fa0-9c57-7829d677ae22.PNG)

And here are the two files in the bucket, the first one should've been deleted. I have to mention that they have the same name, but different extensions.
![ui](https://user-images.githubusercontent.com/44806687/163217982-2141c4fa-872c-4390-8725-17df70d2f3f7.PNG)

**Reply by GaryAustin1** [reply]
200 with [] response is normal for bad filename or RLS versus getting buck a file object in the [].  Your name looks correct in the network info you provided, so I would guess back to RLS.  Did you use just (bucket_id = 'resumes') for both select and delete (versus having the owner check in there also)?  You seem to mention you did at one point.  Also do you have any other policies that don't have a bucket id check in them in other parts of storage policies?

**Reply by kksso** [reply]
I am having the same issue with all correct policies in place I get back a 200 response, no error but the file is not removed from bucket.

**Reply by nilooy** [reply]
same for me, the file is still there, even if i delete from the admin ui, the url remains valid.

**Reply by GaryAustin1** [reply]
Does the url stop being valid after cache period is over?

**Reply by iamthe-observer** [reply]
guys im also facing the same issue so if someone could send code of how they wrote the policies id be grateful

**Reply by bennik88** [reply]
Hey @iamthe-observer I don't know if you already solved this I just stumbled across this discussion and thought I help all upcoming users and you. Below is an example to enable delete (or from the docs `remove()`) files from storage:

1. Log into Supabase Dashboard.
2. Go to **Storage**
3. Go to **Policies**
4. Click **New policy**
5. In this case click on **For full customization** but this is up to you, Supabase will give you some examples when clicking on **Get started quickly**
6. Name your policy e.g. "Enable delete for authenticated users for their own profile image.
7. Select allowed operation **DELETE**
8. And in policy definition use the example snipped below. See the notes below for more information

The below code works with the example folder path and filename `${userId}/${userId}_profileImage.JPEG`

```sql
-- Make sure the policy applies to the avatars bucket only
bucket_id = 'avatars' AND
--  Make sure they can only access their own folder within that bucket
(storage.foldername(name))[1] = auth.uid()::text AND
-- Check the file name of the file and only allow delete if it matches the name below
storage.filename(name) = auth.uid() || '_profileImage.JPEG'
```
Have fun deleting files ;)

**Reply by iamthe-observer** [reply]
> Hey @iamthe-observer I don't know if you already solved this I just stumbled across this discussion and thought I help all upcoming users and you. Below is an example to enable delete (or from the docs `remove()`) files from storage:
> 
> 
> 
> 1. Log into Supabase Dashboard.
> 
> 2. Go to **Storage**
> 
> 3. Go to **Policies**
> 
> 4. Click **New policy**
> 
> 5. In this case click on **For full customization** but this is up to you, Supabase will give you some examples when clicking on **Get started quickly**
> 
> 6. Name your policy e.g. "Enable delete for authenticated users for their own profile image.
> 
> 7. Select allowed operation **DELETE**
> 
> 8. And in policy definition use the example snipped below. See the notes below for more information
> 
> 
> 
> The below code works with the example folder path and filename `${userId}/${userId}_profileImage.JPEG`
> 
> 
> 
> ```sql
> 
> -- Make sure the policy applies to the avatars bucket only
> 
> bucket_id = 'avatars' AND
> 
> --  Make sure they can only access their own folder within that bucket
> 
> (storage.foldername(name))[1] = auth.uid()::text AND
> 
> -- Check the file name of the file and only allow delete if it matches the name below
> 
> storage.filename(name) = auth.uid() || '_profileImage.JPEG'
> 
> ```
> 
> Have fun deleting files ;)
> 
> 

You're a savior from the future thanks. But I solved the issue a while back. Thanks for feedback tho. Im learning a lot from the method you used for the policies.

**Reply by bennik88** [reply]
You're welcome.

**Reply by adamzerner** [reply]
Nice, thanks @bennik88 . To others, note that you might have different requirements (see the code comments).

**Comment by fimalcicd** [comment]
while deleting this error showing 'List<FileObject>' is not a subtype of type 'String' in flutter

**Comment by sturatcliffe** [comment]
Hi, I'm getting this exact issue when trying to delete files from a node.js environment, but I have the correct policies in place - has anyone else had this issue for reasons other than missing RLS policies?

**Reply by GaryAustin1** [reply]
Does it work with RLS true with no user role selected (you have to meet delete and select policy)?  
You don't provide any info on your policies or environment.
You might also try Supabase's Discord for more interactive help.

**Reply by sturatcliffe** [reply]
So my original policy was setup to allow the anon role full CRUD access on an "uploads" bucket, and that was failing, so I tried opening the policy up so it didn't specify a role at all, and it still does not work. Here are a few screenshots of my policies:

![image](https://user-images.githubusercontent.com/17248652/219983284-935057fc-a9ea-4acd-b496-3aee3e4dc3b1.png)

![image](https://user-images.githubusercontent.com/17248652/219983305-ed42b9fa-a717-4bf1-8705-414dfa714fb5.png)

All 4 policies have the same condition applied.

Environment wise the code to delete a file is firing from a server-side action in a Remix application, using version 2.7.1 of the @supabase/supabase-js library.

**Reply by GaryAustin1** [reply]
Any errors that you can tell?  Possibly check Postgres logs around that time.
Your RLS looks OK if that is the same for all and your bucket is lower case uploads...
Calling the correct instance?

Edit:  Guess the pathname being correct is also something to check.  I've seen the wrong path/url being used several times as the issue.

**Reply by sturatcliffe** [reply]
Yes, correct instance is being called.

I'm wondering if it is something path related, but not whether I'm sending the correct path or not. Removals are working for files which are only 1 folder deep, e.g. folder/file.jpg.

However, the ones which are failing are several folders deep - could this have an impact?

**Reply by GaryAustin1** [reply]
Not sure how with the RLS you show.  Certainly if you involve "folders" in the RLS.
Folders are not real constructs.  They are just parts of the path name separated by `/` .  The dashboard UI makes it look like there are folders, but to the API it is just a string.  
You might look in the storage.objects table for a file not being deleted and see if you see anything strange.
I still encourage you to participate in Discord as every post we make here goes to all the former participants from over a year ago.

**Comment by RilDev** [comment]
Hi there,

I was having a similar issue with this code:

```
const { data, error } = await supabase.storage
  .from("my-bucket")
  .remove([
    "/filepath/filename.ext",
  ]);
```

It was returning me `{data: [], error: null}`.

The problem was the initial `/` in the filepath. So I removed it like so: `"/filepath/filename.ext"` -> `"filepath/filename.ext"` and it worked as expected!

**Reply by ozasadnyy** [reply]
Thank you, @RilDev, it worked 🙌

**Comment by evanfeliza** [comment]
Hi current experience this issue its returning `[]` upon checking on logs the delete was accepted

see screenshot below:
![image](https://github.com/supabase/supabase/assets/93525603/47692fcf-478b-4092-955f-f27a2f1b14ee)

Also do i really need to create a folder inside the bucket or not? 

my implementation was directly saving files on a bucket

**Reply by GaryAustin1** [reply]
Folders are not real things.  They are just processed using the path names between /.  So no you don't create folders.

What delete operation are you doing?  You have to use the Storage API and not SQL delete of the row in storage, and you need to meet RLS for select and delete.

**Comment by vincentvella** [comment]
If you're still having issues remvoing images (I know, 3 years later, we're still facing issues), it appears that regardless of whether or not the bucket is public, you need to have a policy set for **SELECT** - it can be something super lenient if you want other users to be able to view other users' profile pictures.
Example:
`(bucket_id = 'profile_pictures'::text)`

**Reply by jeromevvb** [reply]
Thank you! I couldn’t figure out why my file wasn’t being deleted even though the request was successful, but it turns out I needed the SELECT policy.

**Reply by eduardomessiaskulka** [reply]
Thank you so much @vincentvella!!! I spent nearly two hours struggling with this issue, and your tip truly saved me.

**Comment by AadarshaKhatri** [comment]
I had a similar issue (already had the policies updated) but then I was trying to delete the file using the encoded file path, like "Profile%20Image", which wasn' the correct file path. Upon using `decodeURIComponent()`, I was able to decode the file path correctly, resolving the issue and allowing the files to be successfully deleted. I hope this information is helpful!
